modules/ethereum/src/import.rs

```diff
@@ -140,15 +127,19 @@ pub fn import_header<S: Storage>(
 		finalized_blocks.votes,
 	));
 
-	// now mark finalized headers && prune old headers
-	storage.finalize_headers(
-		finalized_blocks.finalized_headers.last().map(|(id, _)| *id),
-		match is_best {
-			true => header_number.checked_sub(prune_depth),
-			false => None,
-		},
+	// compute upper border of updated pruning range
+	let new_best_block_id = if is_best { header_id } else { best_id };
```

Comment:
if the imported header is not new best (i.e. `is_best == false`) then we don't really need to run the whole code at all (pruning & finalizing), do we?

We can assume that it was already run when `best_id` was imported.

Reply:
That's a very good point! But while I hope you're right, I'm unsure about this - will need to check it twice. My concern is whether situation like that is possible:
1) we need 2-of-3 headers for finality;
2) unfinalized headerchain is `H10, H11, H12, H13` - all headers are signed by signer1;
3) signer2 broadcasts the block `H12'` (build on top of `H11`) for some old step.

This header is not the best, because iiuc difficulty is determined by total number of steps in the chain. But at the same time it finalizes headers `H10` and `H11`. I'm not sure if current module impl would reject, or allow import of `H12'` and how it must behave actually.

Reply:
Hmm, interesting. Worth making a test case out of your example to make sure it behaves as we intend.
I assume that if we don't finalize & prune nothing really bad would happen, eventually a chain built on top of `H12'` should take over the `H13` one and then we would finalize & prune `H10` & `H11`
To be honest I think I'd prefer that behavior as more predictable for the users, even though technically `H10`&`H11` can be considered finalized earlier. The worst thing that may happen is that submitting proofs is delayed because of that, but I don't think that extra pruning we have to do when `H13'` gets imported is opening an attack vector.

Reply:
OK, so imo we shouldn't implement this ^^^ behavior. Reasons:
1) if header H has signaled validators change and this block is finalized at block H+N, then starting from block H+N+1 we start verifying headers using new validators set. So we can't miss the moment when block B has been finalized even for non-best chain (i.e. we should track finality on non-best forks);
2) since we don't have any timestamp-related checks now (adding #51 here for reference), malicious authority can maintain competing fork by continuously generating headers => piling up steps => making this fork best => we never finalize anything.

I'll add a test for this case though.

Reply:
I'm convinced :) Thanks!

Reply:
Test is added. Another round of thinking about your comment has resulted in #137 (will try to fix asap) and https://github.com/paritytech/parity-bridges-common/issues/51#issuecomment-644744314 .